src/olympia/amo/middleware.py

```diff
@@ -341,3 +343,25 @@ def __call__(self, request):
         elif max_age_from_response is None:
             patch_cache_control(response, s_maxage=0)
         return response
+
+
+class TokenValidMiddleware:
+    """Middleware to check the FxA auth tokens haven't expired, and refresh if
+    necessary."""
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if (
+            not settings.USE_FAKE_FXA_AUTH
+            and settings.VERIFY_FXA_ACCESS_TOKEN_WEB
+            and request.user.is_authenticated
```

Comment:
This would add, for each non-API request coming with a session cookie, even if the view doesn't need auth at all:
- A database query to load the `UserProfile` (because `request.user` is accessed)
- A database query to load the token later in `fxa_access_token_is_valid()`

`Test500::test_500_renders_correctly_with_no_queries_or_auth_even_when_logged_in` only passes because `VERIFY_FXA_ACCESS_TOKEN_WEB` has been set to `False` in `settings_test.py` (and technically `USE_FAKE_FXA_AUTH` since it's set to `True` in `settings.py` which `settings_test.py` inherits from).

It might be ok now, because we don't have legacy frontend anymore so really the only user-facing, non-api, anonymous pages we have that still matter are devhub logged out and maybe application versions pages, but that seems fragile.

Ideally the fix would be to find a way to stay lazy up until the point where we need the user...

The API equivalent code below might suffer from the same problem, I'd have to check how DRF handles that.

Reply:
> This would add, for each non-API request coming with a session cookie, even if the view doesn't need auth at all:
> 
>     * A database query to load the `UserProfile` (because `request.user` is accessed)

In the general case, it's not going to add a database query because we always check for authenticated status to render the login/logout links, even on views that don't require authentication.
 
>     * A database query to load the token later in `fxa_access_token_is_valid()`

mmm, yes.  But what's the ask here - store the expiry date in the cookie too?  (could that date be changed client side to avoid the check?)

> `Test500::test_500_renders_correctly_with_no_queries_or_auth_even_when_logged_in` only passes because `VERIFY_FXA_ACCESS_TOKEN_WEB` has been set to `False` in `settings_test.py` (and technically `USE_FAKE_FXA_AUTH` since it's set to `True` in `settings.py` which `settings_test.py` inherits from).
> 
> It might be ok now, because we don't have legacy frontend anymore so really the only user-facing, non-api, anonymous pages we have that still matter are devhub logged out and maybe application versions pages, but that seems fragile.

I don't know of any pages other than the 500 that doesn't check if the user is authenticated. (A search for `user_login_include` shows that no other template overrides that block to not show the login/logout links).

> Ideally the fix would be to find a way to stay lazy up until the point where we need the user...

If we don't do a db query for the `FxaToken` instance then we wouldn't need the user pk either, I suppose.  Otherwise, I can't think of an alternative way to first confirm the user is authenticated, and second know the user pk.

> The API equivalent code below might suffer from the same problem, I'd have to check how DRF handles that.

The API code is reusing the `user` variable.